src/Microsoft.OData.Client/ALinq/ResourceBinder.GroupByKeySelectorAnalyzer.cs

```diff
@@ -0,0 +1,114 @@
+            {
+                this.input = source;
+                this.lambdaParameter = paramExpr;
+            }
+
+            /// <summary>
+            /// Analyzes a GroupBy key selector for property or properties that the input sequence is grouped by. 
+            /// </summary>
+            /// <param name="input">The input resource expression.</param>
+            /// <param name="keySelector">Key selector expression to analyze.</param>
+            internal static void Analyze(QueryableResourceExpression input, LambdaExpression keySelector)
+            {
+                Debug.Assert(input != null, "input != null");
+                Debug.Assert(keySelector != null, "keySelector != null");
+
+                GroupByKeySelectorAnalyzer analyzer = new GroupByKeySelectorAnalyzer(input, keySelector.Parameters[0]);
+
+                MemberInitExpression memberInitExpr = keySelector.Body as MemberInitExpression;
+
+                if (memberInitExpr != null)
+                {
+                    analyzer.Visit(memberInitExpr);
+                }
+                else
+                {
+                    analyzer.Visit(keySelector.Body);
+                }
+            }
+
+            /// <inheritdoc/>
+            internal override NewExpression VisitNew(NewExpression nex)
+            {
+                // Maintain a mapping of grouping expression and respective member
+                // The mapping is cross-referenced if any of the grouping expression 
+                // is referenced in result selector
+                for (int i = 0; i < nex.Arguments.Count; i++)
+                {
+                    this.input.Apply.GroupingExpressionsMap.Add(nex.Members[i].Name, nex.Arguments[i]);
+                }
+
+                return base.VisitNew(nex);
+            }
+
+            /// <inheritdoc/>
+            internal override MemberAssignment VisitMemberAssignment(MemberAssignment assignment)
+            {
+                // Maintain a mapping of grouping expression and respective member
+                // The mapping is cross-referenced if any of the grouping expression 
+                // is referenced in result selector
+                this.input.Apply.GroupingExpressionsMap.Add(assignment.Member.Name, assignment.Expression);
+
+                return base.VisitMemberAssignment(assignment);
+            }
+
+            /// <inheritdoc/>
+            internal override Expression VisitMemberAccess(MemberExpression m)
+            {
+                MemberExpression memberExpr = StripTo<MemberExpression>(m);
+
+                // Validate the grouping expression
+                ValidationRules.ValidateGroupingExpression(memberExpr);
+
+                List<ResourceExpression> referencedInputs = new List<ResourceExpression>();
+                Expression boundExpression = InputBinder.Bind(memberExpr, this.input, this.lambdaParameter, referencedInputs);

```

Comment:
@KenitoInc Because of how complex the expressions we're analyzing can be (e.g. the resource set can enclosed in an anonymously-typed transparent scope produced by a `SelectMany` operation), we call this method to confirm that the grouping expression (e.g. `d1.Product.Category.Name`) can be bound to the resource set expression (e.g. `ResourceSetExpression{"Sales"}`)